packages/server/lib/controllers/connection/connectionId/getConnection.ts

```diff
@@ -0,0 +1,114 @@
+        return;
+    }
+
+    const { environment, account } = res.locals;
+
+    const queryParams: GetPublicConnection['Querystring'] = queryParamValues.data;
+    const params: GetPublicConnection['Params'] = paramValue.data;
+
+    const { provider_config_key: providerConfigKey, force_refresh: instantRefresh, refresh_token: returnRefreshToken } = queryParams;
+    const { connectionId } = params;
+
+    const isSync = req.headers['Nango-Is-Sync'] === 'true';
+
+    if (!isSync) {
+        metrics.increment(metrics.Types.GET_CONNECTION, 1, { accountId: account.id });
+    }
+
+    const integration = await configService.getProviderConfig(providerConfigKey, environment.id);
+    if (!integration) {
+        res.status(400).send({ error: { code: 'unknown_provider_config', message: 'Provider does not exists' } });
+        return;
+    }
+
+    const connectionRes = await connectionService.getConnection(connectionId, providerConfigKey, environment.id);
+    if (connectionRes.error || !connectionRes.response) {
+        res.status(404).send({ error: { code: 'not_found', message: 'Failed to find connection' } });
+        return;
+    }
+
+    const credentialResponse = await connectionService.refreshOrTestCredentials({
+        account,
+        environment,
+        connection: connectionRes.response,
+        integration,
+        logContextGetter,
+        instantRefresh: instantRefresh ?? false,
+        onRefreshSuccess: connectionRefreshSuccessHook,
+        onRefreshFailed: connectionRefreshFailedHook
+    });
+    if (credentialResponse.isErr()) {
+        res.status(credentialResponse.error.status).send({
+            error: { code: 'server_error', message: credentialResponse.error.message || 'Failed to refresh or test credentials' }
+        });
+        return;
+    }
+
+    const { value: connection } = credentialResponse;
+
+    if (connection && connection.credentials && connection.credentials.type === 'OAUTH2' && !returnRefreshToken) {
+        if (connection.credentials.refresh_token) {
+            delete connection.credentials.refresh_token;
+        }
+
+        if (connection.credentials.raw && connection.credentials.raw['refresh_token']) {
+            const rawCreds = { ...connection.credentials.raw }; // Properties from 'raw' are not mutable so we need to create a new object.
+            // eslint-disable-next-line @typescript-eslint/no-dynamic-delete
+
+            const { refresh_token, ...rest } = rawCreds;
+            connection.credentials.raw = rest;
+        }
+    }
+
+    // We get connection one last time to get endUser, errors
+    // This is very unoptimized unfortunately
+    const finalConnections = await connectionService.listConnections({ environmentId: environment.id, connectionId, integrationIds: [providerConfigKey] });
+    if (finalConnections.length !== 1 || !finalConnections[0]) {
+        serverError(res, { code: 'server_error', message: 'Failed to get connection' });
```

Comment:
🍖 usage of new serverError function for wrapping a 500 with a trace error and a log message